core/jobs_test.py

```diff
@@ -949,6 +1176,106 @@ def test_events_coming_in_while_batch_job_is_running(self):
+            continuous_computation_data['computation_type'],
+            'StartExplorationEventCounter')
+        self.assertEqual(
+            continuous_computation_data['status_code'], 'idle')
+        self.assertTrue(continuous_computation_data['is_startable'])
+        self.assertFalse(continuous_computation_data['is_stoppable'])
+
+    def test_failing_continuous_job(self):
+        observed_log_messages = []
+
+        def _mock_logging_function(msg, *args):
+            """Mocks logging.error()."""
+            observed_log_messages.append(msg % args)
+
+        StartExplorationEventCounter.start_computation()
+
+        status = StartExplorationEventCounter.get_status_code()
+        self.assertEqual(
+            status, job_models.CONTINUOUS_COMPUTATION_STATUS_CODE_RUNNING)
+
+        with self.swap(logging, 'error', _mock_logging_function):
+            StartExplorationEventCounter.on_batch_job_failure()
+
+        self._process_pending_tasks()
+        StartExplorationEventCounter.stop_computation('admin_user_id')
+
+        self.assertEqual(
+            observed_log_messages, ['Job StartExplorationEventCounter failed.'])
+
+    def test_cancelling_continuous_job(self):
+        observed_log_messages = []
+
+        def _mock_logging_function(msg, *args):
+            """Mocks logging.error()."""
+            observed_log_messages.append(msg % args)
+
+        StartExplorationEventCounter.start_computation()
+
+        status = StartExplorationEventCounter.get_status_code()
+        self.assertEqual(
+            status, job_models.CONTINUOUS_COMPUTATION_STATUS_CODE_RUNNING)
+
+        with self.swap(logging, 'info', _mock_logging_function):
+            StartExplorationEventCounter.on_batch_job_canceled()
+
+        self._process_pending_tasks()
+        StartExplorationEventCounter.stop_computation('admin_user_id')
+
+        self.assertEqual(
+            observed_log_messages,
+            ['Job StartExplorationEventCounter canceled.'])
+
+    def test_kickoff_batch_job_after_previous_one_ends(self):
+        with self.swap(
+            jobs_registry, 'ALL_CONTINUOUS_COMPUTATION_MANAGERS',
+            self.ALL_CC_MANAGERS_FOR_TESTS
+        ):
+            MockContinuousComputationManager.start_computation()
+            (
+                MockContinuousComputationManager  # pylint: disable=protected-access
+                ._kickoff_batch_job_after_previous_one_ends()
+            )
+            status = MockContinuousComputationManager.get_status_code()
+
+            self.assertEqual(
+                status, job_models.CONTINUOUS_COMPUTATION_STATUS_CODE_RUNNING)
+
+            self._process_pending_tasks()
+            MockContinuousComputationManager.stop_computation('admin_user_id')
+            status = MockContinuousComputationManager.get_status_code()
+
+            self.assertEqual(
+                status, job_models.CONTINUOUS_COMPUTATION_STATUS_CODE_IDLE)
+            self.assertEqual(TIMES_RUN, 1)
```

Comment:
You can deifne an additional method on the class if needed. Don't use a global.

Also, in all your tests, please check pre-conditions. I would expect to see an assertion somewhere before this that TIMES_RUN is 0. Please proofread all existing tests and make sure pre-condition checks have been inserted where needed.

Reply:
done